### PR TITLE
[dynamo] support group=None when rewriting collectives

### PR DESCRIPTION
**Pull request by yifuwang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #120560
* #120181
* __->__ #120118

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

File: torch/_dynamo/variables/functions.py [hunk 1]
```diff
@@ -693,10 +694,18 @@ def call_function(
...
         if "async_op" in kwargs and kwargs["async_op"].as_python_constant():
             unimplemented(
                 f"CollectiveFunctionRewriteVariable can't support async_op=True for {self.fn}"
             )
+        group = kwargs.get("group")
+        if group is None or (
+            isinstance(group, variables.ConstantVariable)
+            and group.as_python_constant() is None
+        ):
+            kwargs["group"] = ProcessGroupVariable(dist.group.WORLD, source=self.source)
```
Comment (yifuwang) [inline torch/_dynamo/variables/functions.py:708]: NOTE: this is wrong. We need to popluate the source and install the guard properly.

Comment (wconstab) [reply torch/_dynamo/variables/functions.py:708]: Yea I guess, world shouldn't change but it's proper to guard.

File: torch/_dynamo/variables/distributed.py [hunk 2]
```diff
@@ -247,3 +249,30 @@ def is_process_group(value):
...
+            AttrSource(
+                base=AttrSource(
+                    base=GlobalSource(global_name="torch"),
+                    member="distributed",
+                    get_static=False,
+                ),
+                member="group",
+                get_static=False,
+            ),
+            member="WORLD",
```
Comment (wconstab) [inline torch/_dynamo/variables/distributed.py:271]: Should we generalize this helper to properly retrieve and guard any PG? Not sure if it'd be useful but seems easy to do.

File: torch/_dynamo/variables/functions.py [hunk 3]
```diff
@@ -691,11 +692,27 @@ def call_function(
...
         # since that's the contract for putting a mapping in `traceable_collective_remaps`
-        if "async_op" in kwargs and kwargs["async_op"].as_python_constant():
+
+        # Merge args and kwargs so positional and keyword args
+        # can be located the same way.
+        signature = inspect.signature(self.fn)
+        arg_names = list(signature.parameters.keys())
+        new_kwargs = {}
+        for i in range(len(arg_names)):
+            if i < len(args):
```
Comment (wconstab) [inline torch/_dynamo/variables/functions.py:702]: Doing the manual args/kwargs handling seems open to subtle errors. Is there any way to make python bind the given args/kwargs to a function for you and then use that as the source of truth?

Comment (yifuwang) [reply torch/_dynamo/variables/functions.py:702]: Good point. I think Signature.bind would do the trick.

File: torch/_dynamo/variables/distributed.py (hunk 2 above)
Comment (yifuwang) [reply torch/_dynamo/variables/distributed.py:271]: Probably not that useful. In non-rewriting cases, the correct source is readily available.

Speaking of which, I wonder if I'm complicating things - maybe simply changing `group` to `group or dist.group.WORLD` here would do the trick lol

https://github.com/pytorch/pytorch/blob/637cf4a3f2cfdd364005681636ca885bdc4d5887/torch/distributed/_functional_collectives.py#L996

Comment (awgu) [reply torch/_dynamo/variables/distributed.py:274]: In my understanding, if the default PG cannot change, then `group or dist.group.WORLD` there would be ok. However, it seems from the discussion [here](https://fburl.com/workplace/2x06lxku) (internal link), re-initializing the default PG is possible, though not common or recommended.

Comment (yifuwang) [reply torch/_dynamo/variables/distributed.py:274]: @awgu oh yea the pg needs to be guarded regardless.

Anyway I looked into the alternative. It turns out that it requires introducing more variable trackers for `torch.distributed` and `GroupMember`. We might want to do it eventually, but right now the approach taken by the PR seems reasonable.

**Review by wconstab — APPROVED** [review]

File: torch/distributed/_functional_collectives.py [hunk 4]
```diff
@@ -1035,10 +1035,20 @@ def all_gather_inplace(
...
+    assert all(
+        t.size(0) == tensor.size(0) for t in tensor_list
+    ), "Remapping variable size all_gather is not yet supported"
+
     output = all_gather_tensor(tensor, 0, group, tag)
-    for dst, src in zip(
-        tensor_list, output.split([t.size(0) for t in tensor_list], dim=0)
-    ):
+
+    # Use aten.slice as instead of aten.split because the latter causes
```
Comment (awgu) [inline torch/distributed/_functional_collectives.py:1044]: nit:
```suggestion
    # Use aten.slice instead of aten.split because the latter causes
```

**Review by awgu — APPROVED** [review]
LGTM!

**Comment by yifuwang** [comment]
Hey @sanchitintel, any reason for closing this PR?

**Comment by yifuwang** [comment]
The PR doesn't seem mergeable anymore due to the previous abnormal revert. Re-opened in https://github.com/pytorch/pytorch/pull/121043

**Pull request closed without merging**